homeassistant/components/stream/hls.py

```diff
@@ -0,0 +1,136 @@
+"""
+Provide functionality to stream HLS.
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/stream/hls
+"""
+import datetime
+
+from aiohttp import web
+
+from .core import StreamView, StreamOutput
+
+
+async def async_setup_hls(hass):
+    """Set up api endpoints."""
+    hass.http.register_view(HlsPlaylistView())
+    hass.http.register_view(HlsSegmentView())
+    return '{}/api/hls/{}/playlist.m3u8'
+
+
+class HlsPlaylistView(StreamView):
+    """Stream view to serve a M3U8 stream."""
+
+    url = r'/api/hls/{token:[a-f0-9]+}/playlist.m3u8'
+    name = 'api:stream:hls:playlist'
+    # cors_allowed = True
+
+    async def handle(self, request, stream, sequence):
+        """Return m3u8 playlist."""
+        hass = request.app['hass']
+        renderer = M3U8Renderer(stream, hass.config.api.base_url)
+        track = stream.add_provider(HlsStreamOutput())
+        stream.start()
+        # Wait for a segment to be ready
+        if not track.segments:
+            await track.recv()
+        headers = {
+            'Access-Control-Allow-Origin': '*',
```

Comment:
Shall we get this from `http.cors`?

Reply:
I tried, and ran into issues with setting `cors_allowed = True`.  Did not get the chance to dig in deeper.  Will attempt to do so.

Reply:
Yes, but as a developer, I still couldn’t figure out which CORS Origins to allow in order for my stream to show up on my chromecast correctly. Do you think the average user will be able to?

The cors config also doesn’t allow *

Reply:
I was under the impression that `cors_allowed = True` would set the`*`.

https://github.com/home-assistant/home-assistant/blob/0f6e0aa355af3e25a0426e768825bef8a3e6fcfd/homeassistant/components/http/cors.py#L42-L47

Reply:
Allowed methods is not the same as allowed origins.  Relevant code below:
https://github.com/home-assistant/home-assistant/blob/b18b1cffff8ef5bdcc791999a73e012353937c2e/homeassistant/components/http/__init__.py#L217
https://github.com/home-assistant/home-assistant/blob/0f6e0aa355af3e25a0426e768825bef8a3e6fcfd/homeassistant/components/http/cors.py#L22

Basically, that method requires origins to be defined in the `http` configuration, and from my initial testing, `*` wasn't a valid option.  So the user has to know which Origin their cast devices would be using.

Reply:
`Access-Control-Allow-Origin: *` should not be used in production system.

I don't have access to a system now, but I believe it should be easy to find out which `orgin' Chromecast or Google Home used in its request header. You can eval the request header in your `StreamView` handler

Reply:
Ok, so I retested with `cors_allowed = True`, and it works just fine.  I must have tried it previously before I got the stream working correctly.

Reply:
Have done some debugging, the `Orgin` sent from Google Home is `https://www.gstatic.com`. We may need add this to default cors, or at least put it in the document.

```
<CIMultiDictProxy(
'Host': '192.168.86.141:8123', 
'Connection': 'keep-alive', 
'Origin': 'https://www.gstatic.com', 
'User-Agent': 'Mozilla/5.0 (Linux; Android) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/70.0.3538.47 Safari/537.36 CrKey/1.36.141422', 
'Accept': '*/*', 
'Accept-Encoding': 'gzip, deflate', 
'Accept-Language': 'en-US,en;q=0.9', 
'CAST-DEVICE-CAPABILITIES': '{"bluetooth_supported":true,"display_supported":true,"hi_res_audio_supported":false,"remote_control_input_supported":false,"touch_input_supported":true}'
)>
```